client.go

```diff
@@ -40,35 +40,67 @@ func init() {
 
 // StatsdClient is a client library to send events to StatsD
 type StatsdClient struct {
-	conn             net.Conn
-	addr             string
-	prefix           string
-	eventStringTpl   string
-	Logger           Logger
-	conn_type        string
-	reconnect_ticker *time.Ticker
+	conn            net.Conn
+	addr            string
+	prefix          string
+	eventStringTpl  string
+	Logger          Logger
+	connType        string
+	reconnectTicker *time.Ticker
 }
 
-// NewStatsdClient - Factory
-func NewStatsdClient(addr string, prefix string) *StatsdClient {
+// ConfigurationFunc is a typedef for a function that configures some aspect of
+// a StatsdClient instance
+type ConfigurationFunc func(*StatsdClient)
```

Comment:
@dcarney Do you know of any articles/docs about this design pattern? I'm not familiar with it.

Reply:
Are there flaws or shortcomings of this pattern? One that I'm thinking of is if you had multiple configurations options that were mutually exclusive or dependent on each other how would that be handled? Seems like this pattern creates a dependency on the order of which configuration functions are passed to the caller.

Reply:
Sure, I suppose that could happen depending on the actual config func implementations you create.  The same thing could potentially happen though, if you were using a config struct instead (picture 2 struct fields with conflicting values).

The good news is, since all of the config funcs are under the control of this package, we can have control over how they interact, since we're the package authors.